[sql-32] accounts: add migration code from kvdb to SQL

ViktorTigerstrom opened the pull request:
This PR introduces the migration logic for transitioning the accounts store from kvdb to SQL.

Note that as of this PR, the migration is not yet triggered by any production code, i.e. only tests execute the migration logic.

Part of https://github.com/lightninglabs/lightning-terminal/issues/917

ellemouton commented on accounts/kv_sql_migration_test.go:
```diff
@@ -0,0 +1,339 @@
+			t.Cleanup(func() {
+				require.NoError(t, kvStore.db.Close())
+			})
+
+			t.Run("Postgres", func(t *testing.T) {
+				migrationTest(t, kvStore, testClock, false)
+			})
+
+			t.Run("SQLite", func(t *testing.T) {
+				migrationTest(t, kvStore, testClock, true)
```
in Lit, we've gone with the build flag approach. So can just use the existing NewTestDB methods no? and then if bbolt backend, just skip the test ?

ViktorTigerstrom replied:
Updated as discussed offline :)

ellemouton commented on accounts/store_sql.go:
```diff
@@ -732,6 +733,79 @@ func (s *SQLStore) StoreLastIndexes(ctx context.Context, addIndex,
 	})
 }
 
+func makeInsertAccountParams(account *OffChainBalanceAccount) (
```
let's keep the migration code in the same file

ellemouton commented on accounts/kv_sql_migration_test.go:
```diff
@@ -0,0 +1,339 @@
+// complete migration test.
+func TestAccountStoreMigration(t *testing.T) {
+	t.Parallel()
+
+	ctxc, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
```
can just do `ctx := context.Background()` for tests

ellemouton commented on accounts/kv_sql_migration_test.go:
```diff
@@ -0,0 +1,339 @@
+
+	// First create a shared Postgres instance so we don't spawn a new
+	// docker container for each test.
+	pgFixture := db.NewTestPgFixture(
+		t, db.DefaultPostgresFixtureLifetime, true,
+	)
+	t.Cleanup(func() {
+		pgFixture.TearDown(t)
+	})
+
+	makeSQLDB := func(t *testing.T, clock *clock.TestClock,
+		sqlite bool) (*SQLStore, *db.TransactionExecutor[SQLQueries]) {
+
+		var store *SQLStore
+
+		if sqlite {
+			store = NewSQLStore(db.NewTestSqliteDB(t).BaseDB, clock)
+		} else {
+			store = NewSQLStore(
+				db.NewTestPostgresDB(t).BaseDB, clock,
+			)
```
see other comment about using the existing infrastructure in LiT that uses build flags instead

ViktorTigerstrom replied:
Thanks, updated!

ellemouton commented on accounts/kv_sql_migration_test.go:
```diff
@@ -0,0 +1,339 @@
+		// After that we compare the migrated indices.
+		kvAddIndex, kvSettleIndex, err := kvStore.LastIndexes(ctxc)
+		if errors.Is(err, ErrNoInvoiceIndexKnown) {
+			// If the db doesn't have any indices, we can't compare
+			// them.
```
if the kvstore didnt populate during test set up though, then we will silently exit here. Rather have a test param like "expectLastIndex" so that we know for sure if we are getting the right result here

ViktorTigerstrom replied:
Ah good catch, I actually forgot to store the indexes when i added invoices, so that caught that :rocket:!

ellemouton commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,211 @@
+		overrideAccountTimeZone(kvAccount)
+		overrideAccountTimeZone(migratedAccount)
+
+		if !reflect.DeepEqual(kvAccount, migratedAccount) {
```
i worry about the direct comparisons here because _i think_ it can fail if the AccountInvoice and AccountPayments maps are large since we cannot guarantee order of the maps. 

So adding to this, i think it would be worth it to have a test that generates a large number of invoices & payments (and maybe even many random accounts) - just so that we can really make sure that we arent missing things here. So basically a test case that is not deterministic and does more of a fuzzing style test

ViktorTigerstrom replied:
Added randomization tests to address this :)!

ellemouton reviewed (commented):
Soooooo excited to see this come through! I couldnt help but take a look even though review not requested - just leaving some drive by comments in the mean time :) 

I'll wait till review requested before looking again :)

ViktorTigerstrom commented on accounts/kv_sql_migration_test.go:
```diff
@@ -320,6 +324,16 @@ func TestAccountStoreMigration(t *testing.T) {
 				require.False(t, known)
 			},
 		},
+		{
+			"randomized accounts",
+			true,
+			randomizeAccounts,
+		},
+		{
+			"rapid randomized accounts",
+			true,
+			rapidRandomizeAccounts,
+		},
```
I added 2 tests here currently, where one uses `rapid` for the randomisation, and one uses the normal `"golang.org/x/exp/rand"` testing semi-randomisation.

The drawback with the `rapid` test, is that we can't really generate too many invoices and and payments per account, as you can't really limit the amount of times the `rapid.Check` runs in golang (i.e. the number of accounts we populate the db with), without limiting the amount of runs the `Check` function would run for the entire `litd` project. I.e. if we limited it for this test, that would impact any other tests in the future that uses the `rapid.Check` function. There is an open issue for this in the rapid lib, see: 
https://github.com/flyingmutant/rapid/issues/38

When running the migration locally, the test migration execution time really goes up if you have a lot of accounts with a lot of payments and invoices, and therefore I opted to not add such a test to not make the `make unit` execution time take too long locally. 
 
Therefore, I also added a normal test that doesn't use `rapid` for randomization, so that test adds up to `1000` invoices and payments. 

In the end though , I can delete one of the tests if you prefer any test over the other :)!

bitromortac replied:
I think I prefer `randomizeAccounts` since it's more concise and performant, great you tested out both approaches :+1:

bitromortac reviewed (commented):
Awesome work :rocket:! Will certainly need another pass, since I haven't reviewed many migration PRs

ViktorTigerstrom commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,286 @@
+		AccountID:      sqlID,
+		Hash:           hash[:],
+		Status:         int16(entry.Status),
+		FullAmountMsat: int64(entry.FullAmount),
```
Yeah I think we have to make a choice here, and for migrations in general:
We can either:
1. Assume that since this is a migration, the data we are migrating must have already been sanity checked and rule checked during the initial insertion into the kvdb, and therefore errors like that shouldn't be possible to occur during the migration.
2. We sanity check and check the rules again for all data during the migration.

I have opted for option 1 here and for upcoming migrations, as I do think that should be the case, and we'd only let the migrations use more resources during the migration if we chose option 2 instead.

But I'd very interested to hear from both of you if you think we should instead go with option 2 here.

bitromortac replied:
Yes, I'm unsure, maybe we should be extra careful with amounts, but we can leave it as is.

ViktorTigerstrom replied:
I agree that there's no big difference between 1 & 2 in ths migration. But there will be a difference for future migrations like `sessions`.

So for example in `sessions`: 
To go with option 2, we'd also check when migrating a session that has linked sessions, that all prior sessions has been revoked. However I don't think that really makes sense as that's the type of rule that should have been checked when the entry was added to the `kvdb`, and would only add extra execution load during the migration IMO. 

Would be interested to hear what both of your opinions is in regards to if we should do such sanity checking or not @bitromortac & @ellemouton!

ViktorTigerstrom commented on accounts/kv_sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+)
+
+// TestAccountStoreMigration tests the migration of account store from a bolt
+// backed to a SQL database. Note that this test does not attempt to be a
+// complete migration test.
```
What I mean here is that we are not really testing that the accounts are fully functional after the migration has been done, nor that any linked sessions etc. work properly. 

That'd be more suitable in an itest instead :)!

ViktorTigerstrom commented on accounts/kv_sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+package accounts
```
Updated :)!

ViktorTigerstrom commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,286 @@
+		if err != nil {
+			return fmt.Errorf("unable to fetch migrated "+
+				"account(%v): %w", kvAccount.ID, err)
+		}
```
Good idea, addressed with the latest push!

ViktorTigerstrom commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,286 @@
+	error) {
+
+	return makeSetAccountIndexParams(indexValue, settleIndexName)
+}
```
Was mainly a way to clean up the call site code, and not have to include the constants there, but I agree in hindsight that that's probably complicating readability more than it helps, so I removed that with the latest push :)

ViktorTigerstrom commented:
Thanks for the reviews! I've addressed feedback and left comments with the second last push. In the last push, I rebased on master to address a merge conflict.

bitromortac reviewed (approved):
LGTM :fire::rocket:

ViktorTigerstrom commented:
Thanks for the review @bitromortac! Will address feedback after next review round :rocket:!

ellemouton reviewed (commented):
looking really great!

> Note that as of this PR, the migration is not yet triggered by any production code, i.e. only tests execute the migration logic.

so reviewing this PR made me realise we should perhaps flip the order of things a bit: basically, i still want to be able to test this PR properly (ie, i want to run master branch - add some accounts, then switch to this PR and see that the accounts get migrated properly). 

So this means that under the `dev` tag, we should first just add an option to point to an existing kvdb folder where we can migrate any existing accounts from (if we spin up in sql mode). Note: this probably means adding a set-up PR before we merge this one. But yeah - i think well worth it so that we can test these individually

lightninglabs-deploy commented:
@viktortigerstrom, remember to re-request review from reviewers when ready

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+	makeSQLDB := func(t *testing.T, clock *clock.TestClock) (*SQLStore,
+		*db.TransactionExecutor[SQLQueries]) {
+
+		testDBStore := NewTestDB(t, clock)
```
Thanks, added :)!

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+			},
+		},
+		{
+			"account with updated expiry",
```
Yeah your right, updated the name :).

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+			},
+		},
+		{
+			"account with updated balance",
```
Yeah I wanted to show the other type of update that you can do to an account, and that this didn't cause any issues to the migration. But you're right that it didn't add any real value, so I went ahead and removed the test :).

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -346,3 +360,227 @@ func TestAccountStoreMigration(t *testing.T) {
+		// randomize balance from 1,000 to 100,000,000
+		balance := lnwire.MilliSatoshi(
+			rand.Int63n(100000000-1000) + 1000,
+		)
```
No not really, think I mainly wanted to just have some kind of balance, as I don't think the actual value should matter?

ViktorTigerstrom commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,272 @@
+		}
+	}
+
+	accountId, err := account.ID.ToInt64()
```
Went with `accountAlias` :)

ViktorTigerstrom commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,272 @@
+			String: account.Label,
+			Valid:  true,
+		}
+	}
```
Nice, TIL.

ViktorTigerstrom commented on accounts/sql_migration.go:
```diff
@@ -0,0 +1,272 @@
+	}, nil
+}
+
+func makeAddAccountInvoiceParams(sqlID int64,
+	hash lntypes.Hash) sqlc.AddAccountInvoiceParams {
+
+	return sqlc.AddAccountInvoiceParams{
+		AccountID: sqlID,
+		Hash:      hash[:],
+	}
+}
+
+func makeUpsertAccountPaymentParams(sqlID int64, hash lntypes.Hash,
```
So I chose to keep these for the following reasons:

1. It's following the same pattern present in the `lnd` migration.
2. Especially for future migrations, having these in a separate function is really helpful for line length. 

I hope that's ok, but else I'll update this.

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+		} else {
+			require.NoError(t, err)
+			require.True(t, expectLastIndex)
+		}
```
With the updates, the else block is needed "here" (as the order has changed).

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+		// compare them to the original.
+		// First we compare the migrated accounts to the original ones.
+		kvAccounts, err := kvStore.Accounts(ctx)
+		require.NoError(t, err)
```
Good point, updated with the latest push :). With the other updates, this is now done outside of the function.

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+
+		for i := 0; i < numAccounts; i++ {
+			assertEqualAccounts(t, kvAccounts[i], sqlAccounts[i])
+		}
```
Good idea! Added :)

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+
+		// After that we compare the migrated indices.
+		kvAddIndex, kvSettleIndex, err := kvStore.LastIndexes(ctx)
+		if errors.Is(err, ErrNoInvoiceIndexKnown) {
```
Addressed in latest push :)

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+	}
+
+	for _, test := range tests {
+		tc := test
```
Hmmm the sources I found seems to indicate that this is necessary, so if you know that something has changed, I'd really be interested to hear more about that 😃? 

I can see that in `lnd`, this pattern is preset as soon as we use `t.Parallel()`, however in `litd` it seems like it varies. So I'm not fully sure if this is needed or not :)?

ellemouton replied:
yeah not needed

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -0,0 +1,348 @@
+				require.NoError(t, kvStore.db.Close())
+			})
+
+			migrationTest(t, kvStore, testClock, tc.expectLastIndex)
```
Agree, thanks :pray:! I updated this, and also took https://github.com/lightninglabs/lightning-terminal/pull/1047#discussion_r2075336651 into consideration. So therefore I changed your draft code to also fetch all `accounts` and `indices` prior to the migration.

ViktorTigerstrom commented:
Re-requested a review from you @bitromortac as quite a bit of the testing code has changed, even though you previously ACKed. I hope that's ok

bitromortac reviewed (approved):
LGTM, the test code looks nicer now :tada:. Tested the migration manually as well, I ran a few commands for accounts (account creation/invoices/payments).

ellemouton commented on accounts/sql_migration_test.go:
```diff
@@ -346,3 +360,228 @@ func TestAccountStoreMigration(t *testing.T) {
+	invoiceCounter := uint64(0)
+
+	rapid.Check(t, func(t *rapid.T) {
+		ctx := context.Background()
```
nit: can init ctx outside of rapid.Check to avoid unnecessary alocation

ellemouton reviewed (approved):
tACK, LGTM!

Couple of comments that came up during testing (but not for this PR):
1) currently we use BIGINT for account alias (8 byte) which becomes an `int64` but then for session alias (4 bytes) we use a blob. should we make these the same type? iirc, we changed the account alias just for the space saving. 
2) when we do the actual migration should we delete the old kvdb files? or perhaps we should leave them for the time being. and then in a future release we can do a clean up once we are more sure things are ok
